src/finn/custom_op/fpgadataflow/rtl/thresholding_rtl.py

```diff
@@ -0,0 +1,794 @@
+        """Returns a configuration dictionary containing axilite write commands
+        in order to program the thresholds into the RTL core during runtime.
+        The default address stride for the weights is 1 byte."""
+
+        # thresholds = model.get_initializer(self.onnx_node.input[1])
+        thresholds = weights
+        num_channels, num_weights_per_channel = thresholds.shape
+
+        weight_addr_boundary = find_next_power_of_2(num_weights_per_channel)
+        # Make sure that the next power of 2 (output) is greater than the input
+        assert weight_addr_boundary >= num_weights_per_channel
+
+        config = {}
+        channel_cntr = 0
+        wdt = self.get_weight_datatype()
+        bw_hexdigit = roundup_to_integer_multiple(wdt.bitwidth(), 4)
+        for channel in thresholds:
+            channel_start_addr = channel_cntr * weight_addr_boundary * address_stride
+            weight_cntr = 0
+            addr = 0
+            for weight in channel:
+                key_name = "{}_{}{}_{}{}".format(
+                    "axilite", "ch", str(channel_cntr), "w", str(weight_cntr)
+                )
+                config[key_name] = (
+                    channel_start_addr + addr,
+                    int(
+                        str(
+                            pack_innermost_dim_as_hex_string(
+                                [weight],
+                                wdt,
+                                bw_hexdigit,
+                            )
+                        ),
+                        0,
+                    ),
+                )
+
+                weight_cntr += 1
+                addr += address_stride
+
+            channel_cntr += 1
+
+        return config
+
+    def generate_params(self, model, path):
+        code_gen_dir = path
+        thresholds = model.get_initializer(self.onnx_node.input[1])
+        mem_mode = self.get_nodeattr("mem_mode")
+        if mem_mode == "const":
+            # save thresholds in thresh.h
+            weight_filename = "{}/thresh.h".format(code_gen_dir)
+            self.make_weight_file(thresholds, "hls_header", weight_filename)
+        elif mem_mode == "decoupled":
+            # save decoupled weights for cppsim
+            weight_filename_sim = "{}/thresholds.npy".format(code_gen_dir)
+            self.make_weight_file(thresholds, "decoupled_npy", weight_filename_sim)
+            # also save weights as Verilog .dat file
+            # This file will be ignored when synthesizing UltraScale memory.
+            weight_filename_rtl = "{}/memblock.dat".format(code_gen_dir)
+            self.make_weight_file(thresholds, "decoupled_verilog_dat", weight_filename_rtl)
+        else:
+            raise Exception(
+                """Please set mem_mode to "const", "decoupled",
+                currently no other parameter value is supported!"""
+            )
+    def make_weight_file(self, weights, weight_file_mode, weight_file_name):
```

Comment:
This function seems to be only used by `generate_params` which itself isn't used, for me this looks like it can also be removed.

Reply:
this is a mistake
`generate_param()` was previously called by `code_generation_ipgen()`. but the[ RTL implementation](https://github.com/Xilinx/finn/blob/89378115f6000439e7fd934fa0455d749420a94d/src/finn/custom_op/fpgadataflow/rtlbackend.py#L56) dose not call this any more. i think it should be changed to call `generate_param()` for completeness. 

for thresholding, `generate_param()` would be a direct call to `make_weight_file()` for completeness. 

I will make this update for you to review.

Reply:
a simpler alternative is to call `make_weight_file()` directly from `generate_hdl()` this avoids any changes to the `rtlbackend`